### PR TITLE
PEP 683: Add a PEP for immortal objects.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
I plan on merging/posting this as soon as I get a review from my co-author.

### CODE REVIEW

**Comment by ericsnowcurrently** [comment]
@eduardo-elizondo

**Comment by AA-Turner** [comment]
You can take 683
A

File: pep-06XX-immortal-objects.rst
Comment (AA-Turner) [inline pep-06XX-immortal-objects.rst]: For rendering it would be better to leave the `<TBD>` fields out whilst still undecided

File: pep-06XX-immortal-objects.rst
Comment (AA-Turner) [inline pep-06XX-immortal-objects.rst]: Minor: You never link / explain what the refcount is, although it may be assumed knowledge for the type of people reading this PEP

File: pep-06XX-immortal-objects.rst
Comment (AA-Turner) [inline pep-06XX-immortal-objects.rst]: Minor: This might be more appropriate for "Rejected Ideas", but you don't have that section currently.

File: pep-06XX-immortal-objects.rst
Comment (AA-Turner) [inline pep-06XX-immortal-objects.rst]: It might be useful have a cite for this

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:4]: For rendering it would be better to leave the `<TBD>` fields out whilst still undecided

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:19]: Minor: You never link / explain what the refcount is, although it may be assumed knowledge for the type of people reading this PEP.

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:88]: Minor: This might be more appropriate for "Rejected Ideas", but you don't have that section currently.

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:97]: It might be useful have a citation/link for this.

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:147]: Please may you change this (and the other footnotes) to inline links.

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:208]: Is this line finished?

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:217]: I'm not sure this section adds much, as it redirects to other sections where you address concerns in-line.

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:297]: This section can be removed.

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:303]: Could you change to something like "The reference implementation is contained in CPython PR #spam` please (and make it an inline link).

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:8]:
```suggestion
Created: 10-Feb-2022
```

Comment (AA-Turner) [reply pep-0683.rst:8]: @ericsnowcurrently this is needed to make the build pass
A

File: pep-0683.rst
Comment (ericsnowcurrently) [inline pep-0683.rst:3]: @eduardo-elizondo, would you rather use a different email?

**Comment by brettcannon** [comment]
@ericsnowcurrently please update `CODEOWNERS` so you get asked to review any PRs for the PEP.

File: pep-0683.rst
Comment (brettcannon) [inline pep-0683.rst:230]: How would this work? Some list of pointers that you can iterate through to free at runtime finalization?

Comment (ericsnowcurrently) [reply pep-0683.rst:230]: I'll have to check with Eddie, but I expect the plan was to walk through the GC objects.

File: pep-0683.rst
Comment (brettcannon) [inline pep-0683.rst:262]: Is it worth tweaking this to always return `sys.maxsize()` for immortal objects so people can test if an object is immortal?

File: pep-0683.rst
Comment (ericsnowcurrently) [reply pep-0683.rst:304]: Why an inline link?

File: pep-0683.rst
Comment (ericsnowcurrently) [reply pep-0683.rst:247]: For now I'd rather folks avoid even thinking about immortal objects and keep this as internal as possible.  We can deal with it later if it looks like it might actually be useful.

File: pep-0683.rst
Comment (AA-Turner) [reply pep-0683.rst]: Detail is in #2130 et al, the precis is that footnotes add an extra step of indirection. 

To keep all links in one place, you could use `target_` (no square brackets) and change 

```
.. [target]
   https://target.example/spam/spam/ham/eggs/and/spam
```

to

```
.. _target
   https://target.example/spam/spam/ham/eggs/and/spam
```

A

File: pep-0683.rst
Comment (ericsnowcurrently) [reply pep-0683.rst:94]: I don't really have any.  However, it's fairly common knowledge on the core team.

File: pep-0683.rst
Comment (ericsnowcurrently) [reply pep-0683.rst:7]: There's no API, no (real) user-facing changes, not much controversy, and not a whole lot to decide.  Basically, it is only internal implementation details.  Normally this wouldn't be a PEP but the relevant info is spread around in several email threads, BPO issues, and GH issues.  @encukou asked for an informational PEP to bring the details together for a bit more clarity.

File: pep-0683.rst
Comment (ericsnowcurrently) [reply pep-0683.rst]: Sorry, I meant is there any problem with spelling out the URL directly?

File: pep-0683.rst
Comment (JelleZijlstra) [reply pep-0683.rst:6]: Interesting, I'd assume that a change to core refcounting semantics that causes a 4% slowdown would be controversial enough to require SC approval, but I'll trust your judgment.

Comment (encukou) [reply pep-0683.rst:6]: I've added it to the SC agenda.

> @encukou asked for an informational PEP

Hm, it might turn out to need a standards track one. Anyway, IMO the text is most important.

Comment (encukou) [reply pep-0683.rst:6]: Please make it a normal standards track PEP. Thanks!

— Petr, *on behalf of the SC*

File: pep-0683.rst
Comment (eduardo-elizondo) [reply pep-0683.rst:218]: @brettcannon at a high level, there are three different kinds of cases that we need to care about when immortalizing objects to make sure that we correctly clean them up:

1) Runtime globals: Py_None, Py_True, Py_False, Small Ints, and types defined by: `PyTypeObject Foo =`. None of these instances are not cleaned up today, thus, the behavior will remain the same even after they are made immortal.

2) Immortalizing non-containers during runtime: Examples of these include interned strings. All of these objects have to be found during runtime shutdown, so they must be tracked somewhere. Fortunately, the biggest use case of this would be interned strings which are already tracked in a python dictionary. We should just refer to those at runtime shutdown to properly clean them up. But yes, the generic strategy is to track and deallocate at shutdown.

3) Immortalizing containers during runtime: Example of these include immortalizing the heap after initialization. To correctly track these instances, we'll leverage the GC's permanent generation by pushing all immortalized containers there. During runtime shutdown, the strategy will be to first let the runtime try to do its best effort of deallocating these instances normally. Most of the module deallocation will now be handled by `pylifecycle.c:finalize_modules` which cleans up the remaining modules as best as we can. It will change which modules are available during `__del__` but that's already defined as undefined behavior by the [docs](https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__). Optionally, we could do some topological disorder to guarantee that user modules will be deallocated first before the stdlib modules. Finally, anything leftover (if any) can be found through the permanent generation gc list which we can clear after `finalize_modules`.

**Comment by ericsnowcurrently** [comment]
@eduardo-elizondo, unless you have any objections, I'm going to merge this first thing tomorrow morning and post to python-dev.

File: pep-0683.rst
Comment (brettcannon) [reply pep-0683.rst:218]: > 1. None of these instances are cleaned up today, thus, the behavior will remain the same even after they are made immortal.

Does it make sense to continue to do that, or should we fix the situation?

Comment (ericsnowcurrently) [reply pep-0683.rst:218]: Static objects should not be freed.

**Review by AA-Turner — APPROVED** [review]

File: pep-0683.rst
Comment (AA-Turner) [inline pep-0683.rst:318]: @ericsnowcurrently I noticed on my last read through that this footnote isn't referenced anywhere anymore. If intentional, I'd suggest making it body text in the references section and not a footnote.

A

File: pep-0683.rst
Comment (eduardo-elizondo) [inline pep-0683.rst:21]: `Py_SET_REFCNT`  won't change the refcount either! https://github.com/python/cpython/pull/19474/files#r776528822

File: pep-0683.rst
Comment (eduardo-elizondo) [inline pep-0683.rst:147]: Probably not needed to include here but these are four different strategies that I'll try:
* Immortalized Interned Strings (including Py_IDENTIFIERs)
* Immortalizing heap after startup
* GC improvements for large applications
* Immortalizing Module-level globals at import time

File: pep-0683.rst
Comment (eduardo-elizondo) [inline pep-0683.rst:251]: This won't change! Or rather, it will change but not in the expected way.

This is something that I really looked into to make sure that the refleak tooling wouldn't break from this change. `gettotalrefcount ` relies on `_Py_RefTotal` which is updated independently from the object's refcount field. Thus, an object's immortality will not make the value of `gettotalrefcount` be something really large.

Currently, I've placed the `_Py_RefTotal` update after the immortalization check. Thus, this value now reflects all the increfs/decrefs that did take place. If we move the check to before the immortalization check, it will now reflect all the attempts to incref and decrefs that took place (but they could have been a no-op). I chose the former since it's a bit unintuitive to have a returned value that's larger than the actual effective operations. This is enough to make the refcount leak tooling to work as is. However, if we choose the latter, it will work even in cases where we can upgrade a once mortal object to be immortal (which is beyond of the scope of any of the current changes in the PR). In this case, `_Py_RefTotal` will always be updated regardless of the the immortalization status of the instance, but the returned value might make less sense as a whole.

The final consideration is that we need to make sure that `_PySet_Dummy` is not set to be immortal, otherwise, it will affect the implementation of `_Py_GetRefTotal`. Perhaps we should just remove the use of `_PySet_Dummy` and fix the gdb plugin in another way?

**Review by eduardo-elizondo — APPROVED** [review]
@ericsnowcurrently I've added a bunch of comments. Up to you if you think we need to update the PEP to reflect what I mention there. Otherwise, looks good to me!

**Comment by ericsnowcurrently** [comment]
@eduardo-elizondo Great!  I'll make some updates and merge it.  Thanks for the help.

File: pep-0683.rst
Comment (ericsnowcurrently) [reply pep-0683.rst]: I've mentioned that `sys.gettotalrefcount()` is unaffected but didn't include any of your explanation.  If you think any of it would be helpful to PEP readers then we can add it in a follow-up.

**Comment by CAM-Gerlach** [comment]
Oops, too late on that change—glad I didn't start a full review. Just remember for the future, though—titles don't have periods in English, and neither do commit message summary lines (which is what the PR title maps to).

File: pep-0683.rst
Comment (CAM-Gerlach) [inline pep-0683.rst:4]: Please update the PEP with a link to the actual thread once you create it, so it is easy for others to find without a lot of guessing and Googling.

Also, there are a smattering of relatively non-critical editing issues (aside from the title containing a comma, which is a bit irregular), but as this was merged just as I got to it, I'm a bit late to the party now so if needed, I can make a separate PR fixing those (though they aren't terribly critical).

**Pull request merged**
